[expo-notifications] Handling notifications

sjchmiela opened the pull request:
# Why

Next `expo-notifications` feature.

# How

- `NotificationsHandlerModule` registers at singleton for new notifications/messages
- for each message it _starts up_ a task which emits an event to JS
- in response to the JS event, delegate responds with the appropriate behavior (eg. `shouldShowAlert: true`)
- the behavior is pushed to native side using `NotificationsHandler.handleNotificationAsync` call
- which directs it to the appropriate task
- task handles the behavior (on iOS calls `completionHandler`, on Android it will show the notification once implemented) and finishes
- if for whatever reason delegate didn't respond in 3 seconds, `onTimeout` is called on task, which emits another event to JS (for debugging purposes) and the task finishes

![excalidraw-202012311929-7](https://user-images.githubusercontent.com/1151041/73078318-3ca14180-3ec2-11ea-9220-c7a2f3c1e558.png)

# Test Plan

Tested manually by sending notifications and logging messages that the scheme works both when the delegate responds and when it does not.

sjchmiela committed:
[expo-notifications] Implement notifications handling on iOS

sjchmiela committed:
[expo-notifications] Implement notifications handling on Android

sjchmiela committed:
[expo-notifications] Implement notifications handling in TS

sjchmiela committed:
[bare-expo] Reinstall pods

esamelson reviewed (approved):
🔥

esamelson commented on packages/expo-notifications/src/NotificationsHandler.ts:
```diff
@@ -0,0 +1,69 @@
+let handleSubscription: Subscription | null = null;
+let handleTimeoutSubscription: Subscription | null = null;
+
+export function setNotificationDelegate(delegate: NotificationDelegate): void {
```
can we call this `setNotificationHandler` (with param `handler: NotificationHandler`) rather than `Delegate`? i like that it's more consistent with the module name and i think is also more idiomatic to JS.

sjchmiela committed:
[test-suite] Add test coverage for NotificationHandler

sjchmiela commented:
- changed `NotificationDelegate` to `NotificationHandler`
- added `| null` to `setNotificationHandler` method to let developers remove the handler
- added proper `Notification` type to handler callbacks (from `NotificationsEmitter.types`)
- fixed `EXNotificationsEmitter` bumping the presentation options to at least `Alert`
- added `test-suite` case to verify whether the handler is called
